zcash_primitives/src/transaction/mod.rs

```diff
@@ -125,12 +126,26 @@ impl TransactionData {
         header
     }
 
-    pub fn freeze(self) -> Transaction {
-        Transaction(self)
+    pub fn freeze(self) -> io::Result<Transaction> {
+        Transaction::from_data(self)
     }
 }
 
 impl Transaction {
+    fn from_data(data: TransactionData) -> io::Result<Self> {
+        let mut tx = Transaction(data, TxId([0; 32]));
```

Comment:
It feels we are going back and forth a bit having to construct the transaction and write it's data to an array before we can set the txid. It would be possible to do this all at once, but maybe it's not worth the trouble.

Reply:
By design, I chose to only allow serialization of complete transactions (so that the transaction format is not used for storing "partially-complete transactions", which greatly complicates the job of the transaction parser). Thus it is necessary either to do what we do here, or to have the transaction serializer be a standalone function that takes all the transaction components as inputs. I don't think the complexity of the latter is worth it; the `txid` field of `Transaction` is private, so we know that only this module (and submodules) will be able to see and/or modify it directly.

Reply:
This could be done by moving the `write<W: Write>(&self, mut writer: W)` function from `Transaction` to `TransactionData`, but I could see this being beyond the scope of this PR and we can revisit it later.